### PR TITLE
Hadar/vecops

### PR DESCRIPTION
**Pull request by HadarIngonyama**
cuda-backend-branch: hadar/vecops

### CODE REVIEW

File: examples/c++/vector-api/.devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Make sure NVIDIA Container Toolkit is installed on your host
```
Comment (yshekel) [inline examples/c++/vector-api/.devcontainer/Dockerfile:1]: remove this file

File: examples/c++/vector-api/.devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (yshekel) [inline examples/c++/vector-api/.devcontainer/devcontainer.json:1]: remove

File: examples/c++/vector-api/example.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
+#include <iostream>
+#include <vector>
+#include <memory>
+
+#include "icicle/runtime.h"
+#include "icicle/api/bn254.h"
+#include "icicle/utils/log.h"
+
+
+// SP: I undertstand this code is auto-generated, but I can't get scrip/gen to work.
```
Comment (yshekel) [inline examples/c++/vector-api/example.cpp:10]: I will fix that - need to use the C++ api and not declare those C apis

File: icicle/backend/cpu/src/field/cpu_vec_ops.cpp [hunk 4]
```diff
@@ -17,15 +20,17 @@ enum VecOperation {
...
   VECTOR_SUM,
   VECTOR_PRODUCT,
   SCALAR_ADD_VEC,
   SCALAR_SUB_VEC,
   SCALAR_MUL_VEC,
-  CONVERT_TO_MONTGOMERY,
-  CONVERT_FROM_MONTGOMERY,
   BIT_REVERSE,
   SLICE,
+  REPLACE_ELEMENTS,
```
Comment (yshekel) [inline icicle/backend/cpu/src/field/cpu_vec_ops.cpp:32]: what is replace elements?

File: icicle/backend/cpu/src/field/cpu_vec_ops.cpp [hunk 5]
```diff
@@ -223,27 +334,41 @@ class VectorOpTask : public TaskBase
...
+  uint32_t m_bit_size;       // use in bitrev operation
+  uint64_t m_stride;         // used to support column batch operations
+  uint64_t m_stride_out;     // used in slice operation
+  T*
+    m_output; // pointer to the output. Can be a vector, scalar pointer, or a matrix pointer in case of replace_elements
+  uint32_t m_log_nof_rows; // log of the number of rows in the matrix, used in replace_elements
+  uint32_t m_log_nof_cols; // log of the number of columns in the matrix, used in replace_elements
+  uint32_t m_nof_rows;     // the number of rows in the matrix, used in out of place matrix transpose
+  uint32_t m_nof_cols;     // the number of columns in the matrix, used in out of place matrix transpose
+  const std::vector<uint64_t>* m_start_indices_in_mat; // Indices used in replace_elements operations
```
Comment (yshekel) [inline icicle/backend/cpu/src/field/cpu_vec_ops.cpp:366]: pointer to vector?? why?

File: icicle/tests/test_curve_api.cpp [hunk 6]
```diff
@@ -189,9 +189,9 @@ TEST_F(CurveApiTest, ecntt)
     ntt_release_domain<scalar_t>();
   };
 
-  // run(s_main_target, out_main.get(), "ecntt", VERBOSE /*=measure*/, 1 /*=iters*/);
-  // run(s_ref_target, out_ref.get(), "ecntt", VERBOSE /*=measure*/, 1 /*=iters*/);
-  // ASSERT_EQ(0, memcmp(out_main.get(), out_ref.get(), N * sizeof(projective_t))); // FIXME
+  run(s_main_target, out_main.get(), "ecntt", VERBOSE /*=measure*/, 1 /*=iters*/);
```
Comment (yshekel) [inline icicle/tests/test_curve_api.cpp:192]: you are not rebased I think. This has changed and note we cannot use memcmp to test projective points

File: icicle/tests/test_field_api.cpp [hunk 7]
```diff
@@ -1,7 +1,10 @@
+#include <cstdint>
 #include <gtest/gtest.h>
 #include <iostream>
 #include "dlfcn.h"
+#include <new>
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:5]: what is this for?

File: icicle/tests/test_field_api.cpp [hunk 8]
```diff
@@ -84,16 +89,24 @@ TYPED_TEST(FieldApiTest, FieldSanityTest)
...
-  FieldApiTest<TypeParam>::random_samples(in_a.get(), N);
-  FieldApiTest<TypeParam>::random_samples(in_b.get(), N);
-
-  auto out_main = std::make_unique<TypeParam[]>(N);
-  auto out_ref = std::make_unique<TypeParam[]>(N);
+  int seed = time(0);
+  srand(seed);
+  ICICLE_LOG_DEBUG << "seed = " << seed;
+  const uint64_t N = 1 << (rand() % 15 + 3);
+  // const uint64_t N = 1 << (3);
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:98]: remove commented code from entire file

File: icicle/tests/test_field_api.cpp [hunk 9]
```diff
@@ -120,41 +135,353 @@ TYPED_TEST(FieldApiTest, vectorOps)
   // run(s_reference_target, out_ref.get(), false /*=measure*/, 16 /*=iters*/);
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:135]: why is that commented out? remove or keep if need

File: icicle/tests/test_field_api.cpp [hunk 10]
```diff
@@ -120,41 +135,353 @@ TYPED_TEST(FieldApiTest, vectorOps)
...
+    // run(s_reference_target, nullptr, VERBOSE /*=measure*/, vector_accumulate_wrapper, "vector accumulate", ITERS);
+  // }
   run(s_main_target, nullptr, VERBOSE /*=measure*/, vector_accumulate_wrapper, "vector accumulate", ITERS);
-  ASSERT_EQ(0, memcmp(in_a.get(), temp_result.get(), N * sizeof(TypeParam)));
 
-  // add
-  run(s_reference_target, out_ref.get(), VERBOSE /*=measure*/, vector_add<TypeParam>, "vector add", ITERS);
-  run(s_main_target, out_main.get(), VERBOSE /*=measure*/, vector_add<TypeParam>, "vector add", ITERS);
-  ASSERT_EQ(0, memcmp(out_main.get(), out_ref.get(), N * sizeof(TypeParam)));
+  // for (int i = 0; i < total_size; i++) {
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:171]: again

File: icicle/tests/test_field_api.cpp [hunk 11]
```diff
@@ -120,41 +135,353 @@ TYPED_TEST(FieldApiTest, vectorOps)
...
+  ASSERT_EQ(0, memcmp(out_main.get(), out_ref.get(), total_size * sizeof(TypeParam)));
+
+  // // accumulate
+  FieldApiTest<TypeParam>::random_samples(in_a.get(), total_size);
+  FieldApiTest<TypeParam>::random_samples(in_b.get(), total_size);
+  // if (!s_is_cuda_registered) {
+    for (int i = 0; i < total_size; i++) {
+      out_ref[i] = in_a[i] + in_b[i];
+    }
+  // } else {
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:166]: again

File: icicle/tests/test_field_api.cpp [hunk 12]
```diff
@@ -120,41 +135,353 @@ TYPED_TEST(FieldApiTest, vectorOps)
...
+{
+  int seed = time(0);
+  srand(seed);
+  ICICLE_LOG_DEBUG << "seed = " << seed;
+  const uint64_t N = 1 << (rand() % 15 + 3);
+  const int batch_size = 1 << (rand() % 5);
+  const bool columns_batch = rand() % 2;
+  const int total_size = N * batch_size;
+
+  // const uint64_t N = 1 << (20);
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:283]: again

File: icicle/tests/test_field_api.cpp [hunk 13]
```diff
@@ -120,41 +135,353 @@ TYPED_TEST(FieldApiTest, vectorOps)
...
+TYPED_TEST(FieldApiTest, scalarVectorOps)
+{
+  int seed = time(0);
+  srand(seed);
+  ICICLE_LOG_DEBUG << "seed = " << seed;
+  const uint64_t N = 1 << (rand() % 15 + 3);
+  const int batch_size = 1 << (rand() % 5);
+  const bool columns_batch = rand() % 2;
+
+  // const uint64_t N = 1 << (4);
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:362]: again

File: icicle/tests/test_field_api.cpp [hunk 14]
```diff
@@ -120,41 +135,353 @@ TYPED_TEST(FieldApiTest, vectorOps)
...
+    1 << (rand() % 8 + 2); // cpu implementation for out of place trancpose also supports sizes wich are not powers of 2
+  const int C =
+    1 << (rand() % 8 + 2); // cpu implementation for out of place trancpose also supports sizes wich are not powers of 2
+  const int batch_size = 1 << (rand() % 4);
+  const bool columns_batch = rand() % 2;
+  const bool is_in_place = s_is_cuda_registered? 0 : rand() % 2; //TODO - fix inplace (Hadar: I'm not sure we should support it)
 
-  auto h_out_main = std::make_unique<TypeParam[]>(R * C);
-  auto h_out_ref = std::make_unique<TypeParam[]>(R * C);
+  // const int R = 4; // cpu implementation for out of place trancpose also supports sizes wich are not powers of 2
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:473]: again..

File: icicle/tests/test_field_api.cpp [hunk 15]
```diff
@@ -191,197 +520,625 @@ TYPED_TEST(FieldApiTest, matrixAPIsAsync)
...
       icicle_stream_synchronize(config.stream);
       icicle_free_async(d_in, config.stream);
       icicle_free_async(d_out, config.stream);
     }
   };
 
-  run(s_reference_target, h_out_ref.get(), VERBOSE /*=measure*/, "transpose", ITERS);
-  run(s_main_target, h_out_main.get(), VERBOSE /*=measure*/, "transpose", ITERS);
-  ASSERT_EQ(0, memcmp(h_out_main.get(), h_out_ref.get(), R * C * sizeof(TypeParam)));
+  // // Option 1: Initialize each input matrix in the batch with the same ascending values
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:530]: again

File: icicle/tests/test_field_api.cpp [hunk 16]
```diff
@@ -120,41 +135,353 @@ TYPED_TEST(FieldApiTest, vectorOps)
...
+  // affect the test
 
-  std::memcpy(initial_in_a.get(), in_a.get(), N * sizeof(TypeParam));
-  run(s_reference_target, nullptr, VERBOSE /*=measure*/, vector_accumulate_wrapper, "vector accumulate", ITERS);
-  std::memcpy(temp_result.get(), in_a.get(), N * sizeof(TypeParam));
-  std::memcpy(in_a.get(), initial_in_a.get(), N * sizeof(TypeParam));
+  // // add
+  FieldApiTest<TypeParam>::random_samples(in_a.get(), total_size);
+  FieldApiTest<TypeParam>::random_samples(in_b.get(), total_size);
+  if (!s_is_cuda_registered) {
```
Comment (yshekel) [inline icicle/tests/test_field_api.cpp:149]: I think we can simplify and assume we always have 2 devices to compare against.
Also mentioning cuda specifically in every test will be quite annoying when we have another backend (like metal).

**Review by yshekel — CHANGES_REQUESTED** [review]
need to either remove commented code or uncomment. It's unclear what is the intention. Should it remain ? is it WIP?

File: docs/docs/icicle/golang-bindings/vec-ops.md [hunk 17]
```diff
@@ -4,8 +4,8 @@
 
 Icicle exposes a number of vector operations which a user can use:
 
-* The VecOps API provides efficient vector operations such as addition, subtraction, and multiplication.
-* MatrixTranspose API allows a user to perform a transpose on a vector representation of a matrix
+* The VecOps API provides efficient vector operations such as addition, subtraction, and multiplication, supporting both single and batched operations.
+* MatrixTranspose API allows a user to perform a transpose on a vector representation of a matrix, with support for batched transpositions.
```
Comment (mickeyasa) [inline docs/docs/icicle/golang-bindings/vec-ops.md:8]: what about the rest of the operations

File: docs/docs/icicle/primitives/vec_ops.md [hunk 18]
```diff
@@ -16,6 +16,8 @@ The `VecOpsConfig` struct is a configuration object used to specify parameters f
 - **`is_b_on_device: bool`**: Indicates whether the second input vector (`b`) is already on the device. If `false`, the vector will be copied from the host to the device. This field is optional.
 - **`is_result_on_device: bool`**: Indicates whether the result should be stored on the device. If `false`, the result will be transferred back to the host.
 - **`is_async: bool`**: Specifies whether the vector operation should be performed asynchronously. When `true`, the operation will not block the CPU, allowing other operations to proceed concurrently. Asynchronous execution requires careful synchronization to ensure data integrity.
+- **`batch_size: int`**: Number of vectors (or operations) to process in a batch. Each vector operation will be performed independently on each batch element.
```
Comment (mickeyasa) [inline docs/docs/icicle/primitives/vec_ops.md:19]: the assumption is that all the vectors are concatenated to 1 vector

File: docs/docs/icicle/rust-bindings/vec-ops.md [hunk 19]
```diff
@@ -1,10 +1,10 @@
 # Vector Operations API
 
-Our vector operations API includes fundamental methods for addition, subtraction, and multiplication of vectors, with support for both host and device memory.
+Our vector operations API includes fundamental methods for addition, subtraction, and multiplication of vectors, with support for both host and device memory, as well as batched operations.
 
 ## Vector Operations Configuration
 
-The `VecOpsConfig` struct encapsulates the settings for vector operations, including device context and operation modes.
+The `VecOpsConfig` struct encapsulates the settings for vector operations, including device context, operation modes, and batching parameters.
```
Comment (mickeyasa) [inline docs/docs/icicle/rust-bindings/vec-ops.md:7]: remove , before the "and"

File: icicle/include/icicle/api/bls12_377.h [hunk 20]
```diff
@@ -10,6 +10,45 @@
 #include "icicle/msm.h"
 #include "icicle/vec_ops.h"
 
+extern "C" bool bls12_377_g2_eq(bls12_377::g2_projective_t* point1, bls12_377::g2_projective_t* point2);
```
Comment (mickeyasa) [inline icicle/include/icicle/api/bls12_377.h:13]: where are these came from?

File: icicle/include/icicle/polynomials/default_backend/default_poly_backend.h [hunk 21]
```diff
@@ -65,7 +65,7 @@ namespace icicle {
       config.is_async = true;
       config.stream = m_stream;
 
-      ICICLE_CHECK(icicle::slice(in_coeffs, offset, stride, out_size, config, out_coeffs));
+      ICICLE_CHECK(icicle::slice(in_coeffs, offset, stride, size, out_size, config, out_coeffs));
```
Comment (mickeyasa) [inline icicle/include/icicle/polynomials/default_backend/default_poly_backend.h:68]: "size" should be "in_size"?

File: icicle_v3/include/icicle/mmcs.h [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+
+
+template <typename T>
+  struct Matrix {
+    T* values;
+    size_t width;
+    size_t height;
+  };
+
+eIcicleError build_mmcs_tree(const Matrix<limb_t>* inputs,
```
Comment (mickeyasa) [inline icicle_v3/include/icicle/mmcs.h:20]: what is this file? we don't support mmcs

File: icicle/include/icicle/polynomials/default_backend/default_poly_backend.h (hunk 21 above)
Comment (yshekel) [reply icicle/include/icicle/polynomials/default_backend/default_poly_backend.h:68]: yes. fixing

File: icicle_v3/include/icicle/mmcs.h (hunk 22 above)
Comment (yshekel) [reply icicle_v3/include/icicle/mmcs.h]: seems like a mistake. I removed it

**Review by mickeyasa — APPROVED** [review]

**Review by yshekel — APPROVED** [review]

**Pull request merged**